File: leetcode/78-subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> list = new ArrayList<>();
        backtrack(list, new ArrayList<Integer>(), nums, 0);
        return list;
    }
    public void backtrack(List<List<Integer>> list, List<Integer> item, int[] nums, int start){
        list.add(new ArrayList<Integer>(item));
        for(int i = start; i < nums.length; i++){
            item.add(nums[i]);
            backtrack(list, item, nums, i+1);
            item.remove(item.size()-1);
        }
    }
}

File: leetcode/78-subsets.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> item;
        if(nums.size() == 0) return res;
        backtrack(res, item, 0, nums);
        return res;
    }
    void backtrack(vector<vector<int>>& res, vector<int>& item, int start, vector<int>& nums) {
        if(item.size() > nums.size())
            return;
        res.push_back(vector<int>(item));
        for(int i = start; i < nums.size(); i++) {
            item.push_back(nums[i]);
            backtrack(res, item, i+1, nums);
            item.pop_back();
        }
    }
};
